### PR TITLE
Attempt to improve the workshop

### PR DESCRIPTION
**Pull request by tcharding**
Some improvements (hopefully :)

Patch 1 adds a dependency that reduces setup requiremens.

Patch 2 does:

- Improve the readme incl. workshop instructions
- Add `help` command
- Add docs to `main.rs`
- Add `lib.rs` with constants
- Add helper functions in `lib.rs`
- Re-order code in `lib.rs` in an effort to improve the dev experience

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,50 +2,89 @@
...
+
+0. Install/configure bitcoind regtest node (see below).
+1. Configure `pico-bitcoin-wallet` (see below).
+2. Read the code in `lib.rs`
+3. Implement the `address` command.
+4. Implement the `scan` command.
+5. Implement the `send` command.
+
+
+Solutions are provide per-step as git branches.
```
Comment (Kixunil) [inline README.md:31]: provided

File: README.md [hunk 2]
```diff
@@ -2,50 +2,89 @@
...
 
-This code is intended for demonstration only.
+TL;DR - run `cargo help` and read `main.rs`.
+
+This code is intended for demonstration purposes only.
 If you read it it should give you a better picture of [`rust-bitcoin`](https://docs.rs/bitcoin) and
 associated crates.
-Note that this is just basic introduction, not a comprehensive developer guide.
-It doesn't teach about security or other important practices - it'd require whole book, maybe even multiple books!
+Note that this is just basic a introduction, not a comprehensive developer guide.
```
Comment (Kixunil) [inline README.md]: a basic

File: README.md [hunk 3]
```diff
@@ -2,50 +2,89 @@
...
+```toml
+bitcoind_uri = "http://localhost:18443"
+bitcoind_username = "admin"
+bitcoind_password = "pass"
+```
+
+### Mine coins
+
+After you have completed step 3 (implementing the `address` command) you can use `bitcoin-cli
+generatetoaddress 101 <address>` to mine some coins. This is the only `bitcoin-cli` command you need
```
Comment (Kixunil) [inline README.md]: To do this correctly the wallet would have to handle coinbase transactions correctly. Which seems too much for a simple wallet, so I prefer mine to a Core wallet and then send.

File: README.md [hunk 4]
```diff
@@ -2,50 +2,89 @@
...
 
 * Note that the last block has number equal to block count, so the traditional `last_height..current_height` misses the last one.
-* Private key pair for Taproot needs to be tweked before signing - call `.tap_tweak()` on it and then convert it using `to_inner`.
+* Private key pair for Taproot needs to be tweaked before signing - call `.tap_tweak()` on it and then convert it using `to_inner`.
 * To obtain a block from bitcoind you need to first get its hash using `get_block_hash`.
-* `store_txos` takes an iterator, don't bother with being clever if you are low on time - just collect to `Vec` and pass it in.
-* Don't bother with coin selection - just always spend all inputs, that'd be the best anyway if you actually reuse addresses.
+* `store_txos` takes an iterator, don't bother being clever if you are low on time - just collect to `Vec` and pass it in.
+* Don't bother with coin selection, you can either just always spend all inputs or just spend the
+  first one you find that is big enough (they'll be 50BTC each remember).
```
Comment (Kixunil) [inline README.md]: Only coinbase inputs will be. So picking first is confusing. Also this drops a comment about spending all inputs with same address being the best approach, but thinking about it it's not always the best and the nuance is too overwhelming.

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
+//! pico-bitcoin-wallet
```
Comment (Kixunil) [inline src/lib.rs]: Why complicate it with adding a lib?

File: src/main.rs [hunk 6]
```diff
@@ -1,46 +1,131 @@
...
-    todo!()
+fn send(_args: std::env::Args) -> Result<()> {
+    todo!("Implement send once you have scan working")
 }
 
-fn send(args: std::env::Args) -> Result<()> {
-    let db = db::Db::open();
-    todo!()
+/// Prints the config and data directory paths.
+fn path() -> Result<()> {
```
Comment (Kixunil) [inline src/main.rs]: Maybe this could be `help` instead?

File: src/main.rs [hunk 7]
```diff
@@ -1,46 +1,131 @@
...
+
+// fn get_taproot_address() -> Result<Address> {}
+// fn get_segwit_v0_address() -> Result<Address> {}
+
+/// Scans the Bitcoin blockchain.
+///
+/// Requests blocks from `bitcoind`, starting at the current block height (`db.get_last_height`) and
+/// stores relevant transaction information in the database.
+///
+/// Call this each time you use `bitcoin-cli generatetoaddress` to mine coins to your address.
```
Comment (Kixunil) [inline src/main.rs:45]: Also each time you send from other wallet.

File: src/main.rs [hunk 8]
```diff
@@ -1,46 +1,131 @@
...
+/// You can use a taproot address if you would like to play with taproot spends or alternatively you
+/// can use a native segwit address. Note that the PSBT signing APIs are slightly different for each.
+fn address() -> Result<()> {
+    todo!("Use one of the lines below")
+    // get_segwit_v0_address()
+    // get_taproot_address()
+}
+
+// fn get_taproot_address() -> Result<Address> {}
+// fn get_segwit_v0_address() -> Result<Address> {}
```
Comment (Kixunil) [inline src/main.rs]: IMO this is more complicated than before.

File: steps.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+Workshop steps
```
Comment (Kixunil) [inline steps.md]: These look identical to those in README.

File: README.md (hunk 3 above)
Comment (tcharding) [reply README.md]: Oh good point, I ran into coinbase tx errors while doing this, had not found a minimal way to document that. I like your idea, will update.

File: README.md (hunk 4 above)
Comment (tcharding) [reply README.md]: Agree on the coinbase stuff (as mentioned above). Agree also that mentioning "best approach" is a can of worms best left closed in an hour long tutorial.

File: src/main.rs (hunk 6 above)
Comment (tcharding) [reply src/main.rs]: Good idea, I'll add the info to the help command also.

File: src/main.rs (hunk 7 above)
Comment (tcharding) [reply src/main.rs:45]: Will fix, thanks.

File: steps.md (hunk 9 above)
Comment (tcharding) [reply steps.md]: Oh woops, yes I wrote it in `steps.md` then copied it into readme, will remove `steps.md`, my bad.

File: README.md (hunk 1 above)
Comment (tcharding) [reply README.md:31]: Looks like you haven't pushed them. Maybe better to wait till closer to the conf anyways so we don't have to go back and change any of the branches.

File: README.md (hunk 2 above)
Comment (tcharding) [reply README.md]: fixed

File: src/lib.rs (hunk 5 above)
Comment (tcharding) [reply src/lib.rs]: I agree the extra file is more complicated but I thought that having all the consts in one easy to see place made it easier to notice them for those that don't read the rest of the code closely (*cough* lazy bastards like me). Also we have to put the path helper functions somewhere and better not to put them in `main.rs` I thought.

File: src/main.rs (hunk 6 above)
Comment (tcharding) [reply src/main.rs]: I removed the function all together and just inlined the logic into the `help` command.

File: src/main.rs (hunk 8 above)
Comment (tcharding) [reply src/main.rs]: Removed leaving just the `address` function.

**Comment by tcharding** [comment]
Perhaps I'll add some docs to `db.rs` and `config.rs` to help super green devs?  (In another PR if at all)

File: src/lib.rs (hunk 5 above)
Comment (Kixunil) [reply src/lib.rs]: We can do all these things without it being `lib.rs`. Just like I did with `config` and `db`.

File: README.md (hunk 1 above)
Comment (Kixunil) [reply README.md:31]: I meant typo. But I didn't write them yet, so maybe leave it out for now?

**Comment by Kixunil** [comment]
Yeah, doc for that would be nice. Especially if it works with `cargo doc --open` which we should recommend.

File: src/lib.rs (hunk 5 above)
Comment (tcharding) [reply src/lib.rs]: No worries, I'll move the contents of lib.rs to `config` and `db`.

Comment (tcharding) [reply src/lib.rs]: ok, I was resistant to this suggestion when I wrote the comment above but it actually turned out nicer - thanks for pushing.

**Pull request closed without merging**